### PR TITLE
[feature] add required_providers to terraform block

### PR DESCRIPTION
**Pull request by ryanking**
Note - this will generate warnings on terraform versions < 0.13.

Move provider versions to the `required_providers` block inside the `terraform` block. This is required for downloading providers from the registry in  terraform 0.13.

Along the way I had to shave a few yaks–

1. Renamed `Plan.Providers` to `Plan.ProviderConfigurations` to make room for `Plan.ProviderVersions`
2. Added support for the sentry provider. Now that provider versions must be in the `terraform.required_providers` block and you can only have one such block, we need to teach fogg about all the providers we use. :(

### Test Plan
* CI
* tested on stp and tfe infra repos

### References

### CODE REVIEW

File: config/v2/config.go [hunk 1]
```diff
@@ -124,6 +124,7 @@ type Providers struct {
 	Okta      *OktaProvider      `yaml:"okta,omitempty"`
 	Snowflake *SnowflakeProvider `yaml:"snowflake,omitempty"`
 	Datadog   *DatadogProvider   `yaml:"datadog,omitempty"`
+	Sentry    *SentryProvider    `yaml:"sentry,omitempty"`
```
Comment (ryanking) [inline config/v2/config.go:127]: Added to support tfe-infra repo.

Now that we need to put providers in `terraform { required_providers { ... } }` *and* you can only have one such block, we will need to add all providers we want to use here :(.

File: templates/component/terraform/fogg.tf.tmpl [hunk 2]
```diff
@@ -43,6 +43,104 @@ terraform {
   required_version = "={{ .TerraformVersion }}"
 
   {{ template "backend" .Backend }}
+
+  required_providers {
```
Comment (ryanking) [inline templates/component/terraform/fogg.tf.tmpl:47]: do refactor this into the plan

**Review by jgadling — APPROVED** [review]

File: plan/plan.go [hunk 3]
```diff
@@ -168,23 +200,27 @@ type BlessProvider struct {
...
+	BaseURL *string `yaml:"base_url,omitempty"`
 }
 
 type TfeProvider struct {
 	Enabled  bool    `yaml:"enabled,omitempty"`
-	Version  *string `yaml:"version,omitempty"`
 	Hostname *string `yaml:"hostname,omitempty"`
 }
 
+type KubernetesProvider struct {
```
Comment (edulop91) [inline plan/plan.go:221]: above looked like this provider can have a hostname configuration, should it be here too?

File: plan/plan.go [hunk 4]
```diff
@@ -539,9 +586,13 @@ func resolveComponentCommon(commons ...v2.Common) ComponentCommon {
 	if oktaConfig != nil {
 		oktaPlan = &OktaProvider{
 			OrgName: *oktaConfig.OrgName,
-			Version: oktaConfig.Version,
 			BaseURL: oktaConfig.BaseURL,
 		}
+
+		providerVersions["okta"] = ProviderVersion{
+			Source:  "oktadeveloper/okta",
```
Comment (edulop91) [inline plan/plan.go:593]: We're currently running on a[ forked version](https://github.com/chanzuckerberg/terraform-provider-okta) of this provider

File: plan/plan.go (hunk 3 above)
Comment (ryanking) [reply plan/plan.go:221]: I was intending to only do enough to support the required_providers part of setting up the kubernetes provider, continuing to rely on hand-writing `provider` blocks for kubernetes for now. I don't see a hostname config, there might have been one in an earlier version of this PR.

File: plan/plan.go (hunk 4 above)
Comment (ryanking) [reply plan/plan.go:593]: erg, is that in the registry?

Comment (edulop91) [reply plan/plan.go:593]: not yet, but I can figure that out

Comment (ryanking) [reply plan/plan.go:593]: I can handle it - this is blocking my work and I think I know how to get it published quickly.

Comment (edulop91) [reply plan/plan.go:593]: ok, thank you!

Comment (ryanking) [reply plan/plan.go:593]: https://registry.terraform.io/providers/chanzuckerberg/okta/latest

**Review by edulop91 — APPROVED** [review]

File: config/v2/config.go [hunk 5]
```diff
@@ -183,12 +189,21 @@ type DatadogProvider struct {
 	Version *string `yaml:"version,omitempty"`
 }
 
+type SentryProvider struct {
+	Version *string `yaml:"version,omitempty"`
+	BaseUrl *string `yaml:"base_url,omitempty"`
```
Comment (cziie) [inline config/v2/config.go]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>struct field `BaseUrl` should be `BaseURL` (golint)

File: config/v2/resolvers.go [hunk 6]
```diff
@@ -307,6 +309,28 @@ func ResolveDatadogProvider(commons ...Common) *DatadogProvider {
...
+
+	version := lastNonNil(SentryProviderVersionGetter, commons...)
+	baseURL := lastNonNil(SentryProviderBaseURLGetter, commons...)
+
+	if version != nil {
+		return &SentryProvider{
+			Version: version,
+			BaseUrl: baseURL,
+		}
+	} else {
```
Comment (cziie) [inline config/v2/resolvers.go]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`if` block ends with a `return` statement, so drop this `else` and outdent its block (golint)

File: config/v2/config.go [hunk 7]
```diff
@@ -428,6 +443,20 @@ func (c *Config) Generate(r *rand.Rand, size int) reflect.Value {
 		return nil
 	}
 
+	randKubernetesProvider := func(r *rand.Rand, s int) *KubernetesProvider {
```
Comment (cziie) [inline config/v2/config.go]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`(*Config).Generate$13` - `s` is unused (unparam)

File: config/v2/config.go [hunk 8]
```diff
@@ -428,6 +443,20 @@ func (c *Config) Generate(r *rand.Rand, size int) reflect.Value {
...
+		if r.Float32() < 0.5 {
+			return &KubernetesProvider{}
+		}
+		return nil
+	}
+
+	randSentryProvider := func(r *rand.Rand, s int) *SentryProvider {
```
Comment (cziie) [inline config/v2/config.go]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`(*Config).Generate$14` - `s` is unused (unparam)

File: config/v2/resolvers.go [hunk 9]
```diff
@@ -339,6 +363,39 @@ func ResolveTfeProvider(commons ...Common) *TfeProvider {
...
+
+			if t.Version != nil {
+				version = t.Version
+			}
+
+			if t.Hostname != nil {
+				hostname = t.Hostname
+			}
+		}
+
```
Comment (cziie) [inline config/v2/resolvers.go]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>unnecessary trailing newline (whitespace)

**Pull request merged**